### ISSUE TITLE
Bridging legacy task types that do not use Provider API into tasks that do

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle-native#893]
Given:
```
plugins {
    id("java")
}

class MyTask extends DefaultTask {
    @InputFile
    RegularFileProperty input = project.objects.fileProperty()

    @OutputFile
    RegularFileProperty output = project.objects.fileProperty()

    @TaskAction
    void generate() {
        output.get().asFile.text = input.singleFile.size()
    }
}

tasks.register("consumer", MyTask) {
    // How?
    // input = tasks.named("jar")??
    output = project.layout.buildDirectory.file("consumer.out")
}
```

The `Jar` task does not expose a `Provider<RegularFile>` for its output.  

Ultimately, `Jar` should be changed to use the Provider API, but if someone wanted to bridge this today, how would they do it?

- You can sort of workaround this by changing the input property into a `FileCollection` and relying on Gradle's conversion of a Task -> TaskOutputs.
- You can create your own provider that returns jar.archivePath and manually add the dependency information to the consuming task.  This breaks if the input changes to another task.

Maybe `flatMap` can do some of the lifting?
```
def jar = tasks.named("jar", Jar)
consumer.input = jar.flatMap { magicFileToRegularFileProvider(it.archivePath) }
```

**Comment by JLLeitschuh** [comment]
@big-guy I'm also thinking about adding a getter method to the base class of the `Jar` task to resolve this.
Wondering if you have any naming suggestions.
`getOutputFileProvider()`?

**Comment by big-guy** [comment]
@JLLeitschuh Yes, I think for the `Jar` task in particular, I think adding something to `AbstractArchiveTask` makes sense.  I think something like this could work:

- `getArchivePath` becomes `@Internal` and covered by a new method.
- Add `Provider<RegularFile> getArchiveFile()` as an `@OutputFile`.
- Add a new field `RegularFileProperty archiveFile`
- In the constructor for `AbstractArchiveTask`, the value of `archiveFile` would be:
```
archiveFile.set(getProject().getLayout().file(getProject().provider(new Callable<File>() {
    @Override
    public File call() throws Exception {
        return getArchivePath();
    }
})));
```
- And then a integ test that shows you can depend on `jar.archiveFile` as a task dependency, changes to `archivePath` are reflected with `archiveFile`. 
- And in the task action, where we reference `getArchivePath`, this should probably be `archiveFile`.

**Comment by JLLeitschuh** [comment]
This sounds pretty much like what I was thinking too.
Would I remove the `getArchivePath`method? That would break a lot of plugins.
I like the `getArchiveFile()` name.

**Comment by big-guy** [comment]
No, we can't remove it.  We might be able to deprecate it or at least discourage it and suggest `getArchiveFile` instead.

**Comment by JLLeitschuh** [comment]
Is there a particular reason you're saying `Provider<RegularFile> getArchiveFile()` instead of `RegularFileProperty getArchiveFile()`?

**Comment by JLLeitschuh** [comment]
Okay, I understand. It's strange, but I get it.
